engine/buildkit/executor_spec.go

```diff
@@ -626,114 +616,136 @@ func (w *Worker) setupStdio(_ context.Context, state *execState) error {
+	state.procInfo.Stderr = nopCloser{io.MultiWriter(logs.Stderr, state.procInfo.Stderr)}
+	state.cleanups.add("close logs", logs.Close)
 
-	if strings.HasPrefix(state.otelEndpoint, "/") {
-		// Buildkit currently sets this to /dev/otel-grpc.sock which is not a valid
-		// endpoint URL despite being set in an OTEL_* env var.
-		state.otelEndpoint = "unix://" + state.otelEndpoint
+	listener, err := runInNetNS(ctx, state, func() (net.Listener, error) {
+		return net.Listen("tcp", "127.0.0.1:0")
+	})
+	if err != nil {
+		return fmt.Errorf("otel tcp proxy listen: %w", err)
 	}
 
-	if strings.HasPrefix(state.otelEndpoint, "unix://") && state.procInfo.Meta.NetMode == pb.NetMode_UNSET {
-		// if otel endpoint is unix and we are in CNI mode, setup tcp proxying of the unix endpoint for better client compatibility
-		if state.otelUnixSockSrcPath == "" {
-			return fmt.Errorf("no mount found for otel unix socket %s", state.otelUnixSockSrcPath)
-		}
+	proxyConnPool := pool.New()
+	listenerCtx, cancelListener := context.WithCancel(ctx)
+	listenerPool := pool.New().WithContext(listenerCtx).WithCancelOnError()
 
-		listener, err := runInNetNS(ctx, state, func() (net.Listener, error) {
-			return net.Listen("tcp", "127.0.0.1:0")
-		})
-		if err != nil {
-			return fmt.Errorf("otel tcp proxy listen: %w", err)
-		}
-		state.otelEndpoint = "http://" + listener.Addr().String()
+	otelProto := "http/protobuf"
+	otelEndpoint := "http://" + listener.Addr().String()
+	otelSrv := &http.Server{
+		Handler: w.telemetryPubSub,
 
-		proxyConnPool := pool.New()
-		listenerCtx, cancelListener := context.WithCancel(ctx)
-		listenerPool := pool.New().WithContext(listenerCtx).WithCancelOnError()
+		ReadHeaderTimeout: 10 * time.Second, // for gocritic
+	}
 
-		listenerPool.Go(func(ctx context.Context) error {
-			<-ctx.Done()
-			err := listener.Close()
+	listenerPool.Go(func(ctx context.Context) error {
+		<-ctx.Done()
+		err := listener.Close()
+		if err != nil {
+			return fmt.Errorf("close otel proxy listener: %w", err)
+		}
+		return nil
+	})
+	listenerPool.Go(func(ctx context.Context) error {
+		for {
+			conn, err := listener.Accept()
 			if err != nil {
-				return fmt.Errorf("close otel proxy listener: %w", err)
-			}
-			return nil
-		})
-		listenerPool.Go(func(ctx context.Context) error {
-			for {
-				conn, err := listener.Accept()
-				if err != nil {
-					if errors.Is(err, net.ErrClosed) {
-						err = nil
-					}
-					return err
-				}
-
-				// TODO:(sipsma) logging that existed before? Was it useful?
-
-				remote, err := net.Dial("unix", state.otelUnixSockSrcPath)
-				if err != nil {
-					conn.Close()
-					return fmt.Errorf("dial otel unix socket: %w", err)
+				if errors.Is(err, net.ErrClosed) {
+					err = nil
 				}
+				return err
+			}
 
-				proxyConnPool.Go(func() {
-					defer remote.Close()
-					io.Copy(remote, conn)
-				})
-
-				proxyConnPool.Go(func() {
-					defer conn.Close()
-					io.Copy(conn, remote)
-				})
+			l := &singleConnListener{
```

Comment:
Do we still need to do this or can we just simplify to do `otelSrv.Serve(listener)`? I think manually calling `Accept` was needed for the proxying but since that's gone it seems like we only need the one "real" listener? May be missing something though

Reply:
Haha yes, this can be way simplified - just pushed a commit which reduces it down to basically just `http.Serve` and `srv.Shutdown`. No idea how it ended up so complicated before, was probably flailing late at night. :joy: